### ISSUE TITLE
Move Location API From TestApp To SDK

### ISSUE THREAD

**Issue by bleege** [mapbox/mapbox-gl-native#1940]
The current Location API is custom built and [layered on top of the SDK in the TestApp](https://github.com/mapbox/mapbox-gl-native/blob/0178c94f6d79891d81468a69673263fa62ebcb9c/android/java/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxgl/testapp/MainActivity.java#L285-L313).  Let's extract this into the SDK itself so that it can be used universally.  Model the [API after Google Maps](https://developers.google.com/android/reference/com/google/android/gms/maps/package-summary).

Related to #756

**Comment by ljbade** [comment]
Same as https://github.com/mapbox/mapbox-gl-native/issues/1719

**Comment by bleege** [comment]
The Google Maps API has 2 key access points:

1. Internal to [GoogleMap](https://developers.google.com/android/reference/com/google/android/gms/location/package-summary)
  * [Deprecated](https://developers.google.com/android/reference/com/google/android/gms/maps/GoogleMap.OnMyLocationChangeListener)
  * Google is pushing [Google Play Services Location API](https://developers.google.com/android/reference/com/google/android/gms/location/package-summary)
1. User can provide a custom [LocationSource](https://developers.google.com/android/reference/com/google/android/gms/maps/LocationSource)
  * Corresponding [LocationSource.OnLocationChangedListener](https://developers.google.com/android/reference/com/google/android/gms/maps/LocationSource.OnLocationChangedListener)

**Comment by bleege** [comment]
Ported Location tech and UserLocation API to GL SDK.  Have also refactored the TestApp to run on it.  Testing has revealed that the TestApp is using it correctly and that Location is being found.  The only issue currently is that the UserLocation marker is not being displayed on the MapView.  Next step is to debug that.

**Comment by bleege** [comment]
Turns out the the UserLocation marker is not showing up because the MapView is a View class and can't be added directly.  Experimented with having MapView extend `ViewGroup` and `FrameLayout` instead while keeping a `SurfaceView` object in the class.  The GL map appears to be running just fine, but still issues remaining while trying to add the UserLocation markr (aka ImageView).  Will keep tinkering.

**Comment by ljbade** [comment]
@bleege Ah interesting. Hadn't thought of that being an issue but makes sense.

Perhaps we need MapView to be a ViewGroup/FrameLayout with a inner GLSurfaceView constructed at run time we keep as a private member variable.

Then just move some of the calls to set up GL to this variable.

Going to need to something like this to be able to handle things like the Mapbox logo overlay.

**Comment by bleege** [comment]
I was able to get `MapView` to extend `ViewGroup` instead of `SurfaceView` so that it'd be possible to layer multiple `View` objects on top of it.  The basic implementation needs to be cleaned up a bit as it's currently struggling with `ImageView` because `ImageView` uses a `FrameLayout` instead of the basic `MapView.LayoutParams`.  I'm not too worried about simplifying the basic implementation this to get it to work though.

For a point of comparison, I looked at the [Raster SDK's drawing of UserLocation](https://github.com/mapbox/mapbox-android-sdk/blob/f3fda6f01d01dab9c0962ccc8e8aef2ac2ccd014/MapboxAndroidSDK/src/main/java/com/mapbox/mapboxsdk/overlay/UserLocationOverlay.java#L157-L223).  It avoids the Android View hierarchy all together and instead relies on direct painting using `Canvas`.  I think it's going to be best to make use of the View hierarchy though as we'll also have the Compass to add too.

**Comment by bleege** [comment]
Refactoring to `FrameLayout` is showing some promise.  Was able to get the UserLocation marker to appear.  I'm now working on making sure it updates fast enough and is controllable by the implementing app.  A few things still to iron out, but it's looking promising.

**Comment by bleege** [comment]
Just got UserLocation Dot!  It just required refactoring the `OnMapChangedListener` support to be able to handle N listeners instead of just one.  Performance seems decent for now in the TestApp.

Next up is to move the Compass to the SDK and then everything should be SDK side, instead of app development side.

![device-2015-08-04-190457](https://cloud.githubusercontent.com/assets/989928/9075741/0ca7e0ac-3ade-11e5-9ea9-3fea3144bacd.png)

**Comment by ljbade** [comment]
Awesome work @bleege. What other UI widgets are missing from the map compared to iOS? Logo? Attribution?

**Comment by bleege** [comment]
Initial port of Compass from TestApp to SDK is done.  Now adding [Google Maps API compatible on/off API](https://developers.google.com/android/reference/com/google/android/gms/maps/UiSettings.html#isCompassEnabled%28%29).

**Comment by bleege** [comment]
Added Google Maps API as well as a way to test it in the TestApp.  All :eyes: good.

Punting on User Dot location accuracy for now, as they're likely be a new image UI / UX for it (ticket coming soon).  We also need the Location API in the SDK sooner rather than later for testing.

Merge is up next.

**Comment by bleege** [comment]
Merged in #1999.

**Comment by bleege** [comment]
Ugh... found a NullPointerException...

```java
    java.lang.NullPointerException: Attempt to invoke virtual method 'void android.widget.ImageView.setVisibility(int)' on a null object reference
            at com.mapbox.mapboxgl.views.MapView.updateMap(MapView.java:1657)
            at com.mapbox.mapboxgl.views.MapView.access$800(MapView.java:65)
            at com.mapbox.mapboxgl.views.MapView$1.onMapChanged(MapView.java:601)
            at com.mapbox.mapboxgl.views.MapView$3.run(MapView.java:1410)
            at android.os.Handler.handleCallback(Handler.java:739)
            at android.os.Handler.dispatchMessage(Handler.java:95)
            at android.os.Looper.loop(Looper.java:145)
            at android.app.ActivityThread.main(ActivityThread.java:5835)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
```

**Comment by bleege** [comment]
Addressed in #2000.  Will roll new SNAPSHOT now.
